beginend.el

```diff
@@ -232,6 +232,32 @@ BEGIN-BODY and END-BODY are two `progn' expressions passed to respectively
   (progn
     (prodigy-last)))
 
+;; Copy/pasted from expand-region
+(defun beginend--point-is-in-comment-p ()
+  "Return non-nil if point is in comment."
+  (or (nth 4 (syntax-ppss))
+      (memq (get-text-property (point) 'face)
+            '(font-lock-comment-face font-lock-comment-delimiter-face))))
+
+(defun beginend--prog-mode-code-position-p ()
+  "Return non-nil if point, at beginning of line, is inside code."
+  (not
+   (or (beginend--point-is-in-comment-p)
+       (eq (point) (line-end-position))
```

Comment:
The code considers the point to be at the beginning of the line already (this is expressed in the comment). The check `(eq (point) (line-end-position))` checks for empty lines. All in all, the function `beginend--prog-mode-code-position-p` considers a line to be code if non of the condition is satisfied:

- the line starts with a comment
- the line is empty
- the line contains ^L (frequently used to separate "pages" of code)